presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleTTLNodeSelector.java

```diff
@@ -0,0 +1,276 @@
+        return selectNodes(limit, new ResettableRandomizedIterator<>(eligibleNodes));
+    }
+
+    @Override
+    public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)
+    {
+        Multimap<InternalNode, Split> assignment = HashMultimap.create();
+        NodeMap nodeMap = this.nodeMap.get().get();
+        NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
+
+        List<InternalNode> eligibleNodes = getEligibleNodes(simpleNodeSelector.getMaxTasksPerStage(), nodeMap, existingTasks);
+        NodeSelection randomNodeSelection = new RandomNodeSelection(eligibleNodes, simpleNodeSelector.getMinCandidates());
+
+        Set<InternalNode> blockedExactNodes = new HashSet<>();
+        boolean splitWaitingForAnyNode = false;
+        int maxPendingSplitsPerTask = simpleNodeSelector.getMaxPendingSplitsPerTask();
+
+        OptionalInt preferredNodeCount = OptionalInt.empty();
+        for (Split split : splits) {
+            List<InternalNode> candidateNodes;
+            if (split.getNodeSelectionStrategy() == NodeSelectionStrategy.NO_PREFERENCE) {
+                candidateNodes = randomNodeSelection.pickNodes(split);
+            }
+            else {
+                throw new PrestoException(NODE_SELECTION_NOT_SUPPORTED, format("Unsupported node selection strategy for TTL scheduling %s", split.getNodeSelectionStrategy()));
+            }
+
+            if (candidateNodes.isEmpty()) {
+                log.warn("No nodes available to schedule %s. Available nodes %s", split, nodeMap.getActiveNodes());
+                throw new PrestoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            Optional<InternalNodeInfo> chosenNodeInfo = simpleNodeSelector.chooseLeastBusyNode(
+                    candidateNodes,
+                    assignmentStats::getTotalSplitCount,
+                    preferredNodeCount,
+                    simpleNodeSelector.getMaxSplitsPerNode(),
+                    assignmentStats);
+            if (!chosenNodeInfo.isPresent()) {
+                chosenNodeInfo = simpleNodeSelector.chooseLeastBusyNode(candidateNodes, assignmentStats::getQueuedSplitCountForStage, preferredNodeCount, maxPendingSplitsPerTask, assignmentStats);
+            }
+
+            if (chosenNodeInfo.isPresent()) {
+                split = new Split(
+                        split.getConnectorId(),
+                        split.getTransactionHandle(),
+                        split.getConnectorSplit(),
+                        split.getLifespan(),
+                        new SplitContext(chosenNodeInfo.get().isCacheable()));
+
+                InternalNode chosenNode = chosenNodeInfo.get().getInternalNode();
+                assignment.put(chosenNode, split);
+                assignmentStats.addAssignedSplit(chosenNode);
+            }
+            else {
+                splitWaitingForAnyNode = true;
```

Comment:
Looks like we can move this into the above if check `if (!chosenNodeInfo.isPresent())`

Reply:
If `chosenNodeInfo` is not present, we try to find it again on line 182, so we'd have to check if `chosenNodeInfo` is present again if we want to move this into the check `if (!chosenNodeInfo.isPresent())`.